File: 5-longest-palindromic-substring/5-longest-palindromic-substring.java
class Solution {
  public String longestPalindrome(String s) {
    String str = "^#" + String.join("#", s.split("")) + "#$";
    int len = str.length();
    int[] dp = new int[len];
    Arrays.fill(dp, 1);
    int index = 0;
    int center = 0;
    int right = 0;
    for (int i = 2; i < len - 2; i++) {
      if (i < right - 1) {
        dp[i] = Math.min(right - i, dp[2*center-i]);
      }
      while (str.charAt(i-dp[i]) == str.charAt(i+dp[i])) {
        dp[i] += 1;
      }
      if (i + dp[i] > right) {
        center = i;
        right = i + dp[i];
      }
      if (dp[i] > dp[index]) {
        index = i;
      }
    }
    int realIndex = (index-dp[index])/2;
    return s.substring(realIndex, realIndex + dp[index]-1);
  }
}

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.scala
object Solution {
  def longestPalindrome(s: String): String = {
    if (s.length <= 1) return s
    val processed = "^#" + s.mkString("#") + "#$"
    val len = processed.length
    val dp = Array.fill[Int](len)(1)
    var maxIndex = 1
    var c = 2
    var r = 3
    var i = 3
    while (i < len - 2) {
      if (i < r - 1) {
        dp(i) = scala.math.min(dp(c*2-i), r - i)
      }
      while (processed(i-dp(i)) == processed(i+dp(i))) {
        dp(i) += 1
      }
      if (i + dp(i) > r) {
        c = i
        r = i + dp(i)
      }
      if (dp(i) > dp(maxIndex)) {
        maxIndex = i
      }
      i += 1
    }
    val index = (maxIndex - dp(maxIndex)) / 2
    return s.substring(index, index + dp(maxIndex)-1)
  }
}